### PR TITLE
refactor: Move repository of changesets to domain from env

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactoring

### Purpose
- Move repository of changesets to domain from env

### CODE REVIEW

File: sphinx/directives/other.py [hunk 1]
```diff
@@ -261,8 +261,8 @@ def run(self):
                                                 classes=['versionmodified']),
                                    translatable=False)
             node.append(para)
-        # XXX should record node.source as well
```
Comment (tk0miya) [inline sphinx/directives/other.py]: Note: This is already done on `set_source_info()` calling above.

**Comment by tk0miya** [comment]
@shimizukawa what do you think?

File: CHANGES [hunk 2]
```diff
@@ -83,9 +83,11 @@ Deprecated
 * ``env._read_parallel()`` is deprecated
 * ``env.write_doctree()`` is deprecated
 * ``env._nitpick_ignore`` is deprecated
+* ``env.versionchanges`` is deprecated
 * ``env.dump()`` is deprecated
 * ``env.dumps()`` is deprecated
 * ``env.topickle()`` is deprecated
+* ``env.note_versionchange()`` is deprecated
```
Comment (shimizukawa) [inline CHANGES:91]: The impact seems to be low

影響は少なそう

File: sphinx/builders/changes.py [hunk 3]
```diff
@@ -60,28 +62,30 @@ def get_outdated_docs(self):
...
             return
         logger.info(bold('writing summary file...'))
-        for type, docname, lineno, module, descname, content in \
-                self.env.versionchanges[version]:
-            if isinstance(descname, tuple):
-                descname = descname[0]
-            ttext = self.typemap[type]
-            context = content.replace('\n', ' ')
-            if descname and docname.startswith('c-api'):
+        for changeset in domain.get_changesets_for(version):
```
Comment (shimizukawa) [inline sphinx/builders/changes.py:73]: It became easier to read.

読みやすくなりました

**Review by shimizukawa — APPROVED** [review]
This refactoring makes it easier to understand how to hold data.
However, unlike other domains, the changeset domain does not provide a namespace of reST notation. For this reason, I feel it is incompatible to call it a domain. 
Ignoring compatibility, it seems good to change the directive as follows.

- `versionadded` -> `version:added`
- `versionchanged` -> `version:changed`
- `deprecated` -> `version:deprecated`

In reality, such a change is not possible due to its large impact, but I think that it may be good to move the implementation code of these three directives into `sphinx/domains/changeset.py` as well.

**Comment by tk0miya** [comment]
@shimizukawa Thank you for reviewing. I'm happy to agree with you.

I know namespace is one of main feature of Sphinx domains. But it is not a requirement. I think the most important responsibility of Sphinx domain is a collection of indices.
Indeed, the `version*` directives don't belong to namespace. But it works as collector of changesets. And changes builder also works as exporter of changesets. I know there are no namespace, cross reference and index page. But I believe this must be a Sphinx domain because it is a collection of changesets.

+1 for moving the implementation of directives under `sphinx.domains.changesets`. So I'll do it later.

**Comment by tk0miya** [comment]
moved.

**Review by shimizukawa — APPROVED** [review]
again, LGTM!

**Comment by tk0miya** [comment]
Thank you for reviewing!

**Pull request merged**